determinator/src/helpers.rs

```diff
@@ -0,0 +1,375 @@
+            .map(|f| Point3f::new(f.x as f32, f.y as f32, f.z as f32))
+            .collect::<Vec<_>>(),
+    ))
+}
+
+fn point_pair_to_correspondence(
+    img_points: Vector<Point_<f32>>,
+    obj_points: Vec<opencv::core::Point3_<f32>>,
+) -> Vec<ImgObjCorrespondence> {
+    img_points
+        .into_iter()
+        .zip(obj_points)
+        .map(|f| ImgObjCorrespondence::new(point3f_to3d(f.1), point2f_to2d(f.0)))
+        .collect()
+}
+
+pub fn ref_keypoints(args: &Args) -> (Vec<KeyPoint>, Option<Vec<Vec<u8>>>) {
+    match args.demo {
+        true => {
+            // TIHI @Murmeldyret, here be no side effects
+            let points: Result<Vec<KeyPoint>, _> = (1..=7)
+                .map(|f| (f, (1..=7)))
+                .flat_map(|row| {
+                    row.1
+                        .map(move |col| KeyPoint::new_coords_def(row.0 as f32, col as f32, 1.0))
+                })
+                .collect();
+            (points.expect("Failed to create keypoints"), None)
+        }
+        false => {
+            let (keypoints_from_db, descriptors) = keypoints_from_db(
+                &env::var("DATABASE_URL").expect("failed to read DATABASE_URL"),
+                args,
+            );
+            (keypoints_from_db, Some(descriptors))
+        }
+    }
+}
+
+fn keypoints_from_db(conn_url: &str, arg: &Args) -> (Vec<KeyPoint>, Vec<Vec<u8>>) {
+    let conn: DbType = Arc::new(Mutex::new(
+        Connection::establish(conn_url).expect("Failed to connect to database"),
+    ));
+
+    // retrieve keypoints from mosaic image
+    let ref_keypoints = feature_database::keypointdb::Keypoint::read_keypoints_from_lod(
+        &mut conn.lock().expect("Mutex poisoning"),
+        arg.lod,
+    )
+    .expect("Failed to query database");
+    // Map keypoints to opencv compatible type
+    let (ref_keypoints, ref_descriptors) = db_kp_to_opencv_kp(ref_keypoints);
+
+    (ref_keypoints, ref_descriptors)
+}
+
+fn db_kp_to_opencv_kp(
+    ref_keypoints: Vec<feature_database::models::Keypoint>,
+) -> (Vec<KeyPoint>, Vec<Vec<u8>>) {
+    let (ref_keypoints, ref_descriptors): (Vec<_>, Vec<_>) = ref_keypoints
+        .into_iter()
+        .map(|f| {
+            (
+                opencv::core::KeyPoint::new_point(
+                    Point2f::new(f.x_coord, f.y_coord),
+                    f.size,
+                    f.angle,
+                    f.response,
+                    f.octave,
+                    f.class_id,
+                )
+                .expect("error in converting db keypoint to opencv keypoint"),
+                f.descriptor,
+            )
+        })
+        .unzip();
+    (ref_keypoints, ref_descriptors)
+}
+
+// TODO: kan godt være topo parameter skal ændres til en anden type
+pub fn get_3d_world_coord_from_2d_point(points: Vec<Point2d>, db: DbType) -> Vec<Point3d> {
+    points
+        .into_iter()
+        .map(|p| {
+            let world_coord: (f64, f64, f64) = todo!("@rasmus"); //@Murmeldyret
```

Comment:
```suggestion
            let world_coord: (f64, f64, f64) = get_world_coordinates(db, p.x, p.y);
```
Det er den her funktion som Rasmus har lavet og burde virke når hans elevation #34 pull request bliver approved

Reply:
jepsen 👍